### PR TITLE
Refactor SpinWick code

### PR DESCRIPTION
**Pull request by gabrieljackson**
This change does the following:
 - Attempts to cleanup commonly-used code and to create helper
   functions where possible.
 - Refactors the SpinWick code in order to return errors up the
   stack which allows for cleaner logging and error handling.
 - Refactors the waitForBuild function in order to pass in a
   context to handle timeouts as well as to allow it to handle
   missing build links without immediately failing.
 - Adds basic tests to the server package
 - Adds circle-ci job to run tests

### CODE REVIEW

**Comment by gabrieljackson** [comment]
This is a fairly large change, so we should probably merge when we have some time to test out SpinWicks right after it is deployed.

File: server/spinwick.go [hunk 1]
```diff
@@ -77,96 +77,97 @@ type Installation struct {
...
+		mlog.Info("PR already has a SpinWick created", mlog.String("repo_name", pr.RepoName), mlog.Int("pr", pr.Number))
+		return "", false, nil
 	}
 
-	credentials, ok := Config.JenkinsCredentials[repo.JenkinsServer]
-	if !ok {
-		mlog.Error("No Jenkins credentials for server required for PR", mlog.String("jenkins", repo.JenkinsServer), mlog.Int("pr", pr.Number), mlog.String("repo_owner", pr.RepoOwner), mlog.String("repo_name", pr.RepoName))
-		commentOnIssue(pr.RepoOwner, pr.RepoName, pr.Number, Config.SetupSpinmintFailedMessage)
-		return fmt.Errorf("No Jenkins credentials for server required for PR")
+	mlog.Info("No SpinWick for this PR in the Database. Will create a new one.")
```
Comment (jwilander) [inline server/spinwick.go]:
```suggestion
	mlog.Info("No SpinWick for this PR in the database. Creating a new one.")
```

File: server/spinwick.go [hunk 2]
```diff
@@ -236,91 +242,102 @@ func updateSpinWick(pr *model.PullRequest) {
...
+		}
+	}
 
-	destroyMMInstallation(instanceClusterID)
-	// Remove from the local db
 	removeSpinmintInfo(instanceClusterID)
 }
 
-func destroyMMInstallation(instanceClusterID string) {
+// destroySpinWick destroys a SpinWick and returns an error as well as a bool
```
Comment (jwilander) [inline server/spinwick.go]:
```suggestion
// destroyMMInstallation destroys a SpinWick and returns an error as well as a bool
```

File: server/spinwick.go [hunk 3]
```diff
@@ -655,12 +671,14 @@ func checkMMPing(client *mattermostModel.Client4, ctx context.Context) error {
...
-func NewInt(n int) *int          { return &n }
-func NewInt64(n int64) *int64    { return &n }
-func NewInt32(n int32) *int32    { return &n }
-func NewString(s string) *string { return &s }
+func isSpinWickLabel(label string) bool {
+	if label == Config.SetupSpinWick || label == Config.SetupSpinWickHA {
+		return true
+	}
+
+	return false
```
Comment (jwilander) [inline server/spinwick.go]: nit: could be simplified to `return Config.SetupSpinWick || label == Config.SetupSpinWickHA`

**Comment by jwilander** [comment]
We can merge and test tomorrow after sprint planning

**Pull request merged**